src/vmm/src/devices/virtio/iovec.rs

```diff
@@ -133,16 +90,70 @@ impl IoVecBuffer {
     /// # Returns
     ///
     /// The number of bytes read (if any)
-    pub fn read_at(&self, buf: &mut [u8], offset: usize) -> Option<usize> {
+    pub fn read_at(&self, mut buf: &mut [u8], offset: usize) -> Option<usize> {
         if offset < self.len() {
             // Make sure we only read up to the end of the `IoVecBuffer`.
             let size = buf.len().min(self.len() - offset);
-            Some(self.read_subregion(&mut buf[..size], offset))
+            // write_volatile for &mut [u8] is infallible
+            self.read_volatile_at(&mut buf, offset, size).ok()
         } else {
             // If `offset` is past size, there's nothing to read.
             None
         }
     }
+
+    /// Reads up to `limit` bytes from the `IoVecBuffer` starting at the given offset.
+    ///
+    /// This will try to write to the given [`WriteVolatile`].
+    pub fn read_volatile_at<W: WriteVolatile>(
+        &self,
+        dst: &mut W,
+        mut offset: usize,
+        mut limit: usize,
+    ) -> Result<usize, VolatileMemoryError> {
+        let mut total_bytes_read = 0;
+
+        for iov in &self.vecs {
+            if limit == 0 {
+                break;
+            }
+
+            if offset >= iov.iov_len {
+                offset -= iov.iov_len;
+                continue;
+            }
+
+            let mut slice =
+                // SAFETY: the constructor IoVecBufferMut::from_descriptor_chain ensures that
+                // all iovecs contained point towards valid ranges of guest memory
+                unsafe { VolatileSlice::new(iov.iov_base.cast(), iov.iov_len).offset(offset)? };
+            offset = 0;
+
+            if slice.len() > limit {
+                slice = slice.subslice(0, limit)?;
+            }
+
+            let bytes_read = loop {
+                match dst.write_volatile(&slice) {
+                    Err(VolatileMemoryError::IOError(err))
+                        if err.kind() == ErrorKind::Interrupted =>
+                    {
+                        continue
+                    }
+                    Ok(bytes_read) => break bytes_read,
+                    Err(volatile_memory_error) => return Err(volatile_memory_error),
+                }
+            };
+            total_bytes_read += bytes_read;
+
+            if bytes_read < slice.len() {
+                break;
+            }
```

Comment:
Mh, the old code only used u8-slices, and for u8 slices if we call this method through `read_at`, we cannot return anything "in-between" (because of the Read/Write semantics of u8 slices). The kani harnesses actually verify this property. The vsock code on the other hand uses UnixStreams as the Read/Write backend, and thus depends on this function not being "exact". Maybe we can add a check for "buffer was filled" to `read_at/write_at` and return `None` otherwise?

Reply:
> Mh, the old code only used u8-slices, and for u8 slices if we call this method through read_at, we cannot return anything "in-between" (because of the Read/Write semantics of u8 slices). 

Yup, that's my point. Before we new that for our use-case we cared/expected only 0 or `length`. That's why we ended up with the awkward API semantics.

Since this is now not the case any more, I think we should fix the API. But maybe it can happen in a follow up PR (?)

Reply:
We've synced on this off-line, I've added a commit at the end of this series that fixes up the API semantics slightly :)